apidef/graphql_config_adapter.go

```diff
@@ -0,0 +1,284 @@
+package apidef
+
+import (
+	"encoding/json"
+	"errors"
+	"net/http"
+
+	"github.com/jensneuse/graphql-go-tools/pkg/ast"
+	"github.com/jensneuse/graphql-go-tools/pkg/astparser"
+	"github.com/jensneuse/graphql-go-tools/pkg/engine/datasource/graphql_datasource"
+	"github.com/jensneuse/graphql-go-tools/pkg/engine/datasource/httpclient"
+	"github.com/jensneuse/graphql-go-tools/pkg/engine/datasource/rest_datasource"
+	"github.com/jensneuse/graphql-go-tools/pkg/engine/plan"
+	"github.com/jensneuse/graphql-go-tools/pkg/graphql"
+)
+
+var ErrUnsupportedGraphQLConfigVersion = errors.New("provided version of GraphQL config is not supported for this operation")
+
+type GraphQLConfigAdapter struct {
+	config     GraphQLConfig
+	httpClient *http.Client
+}
+
+func NewGraphQLConfigAdapter(config GraphQLConfig) GraphQLConfigAdapter {
+	return GraphQLConfigAdapter{config: config}
+}
+
+func (g *GraphQLConfigAdapter) EngineConfigV2() (*graphql.EngineV2Configuration, error) {
+	if g.config.Version != GraphQLConfigVersion2 {
+		return nil, ErrUnsupportedGraphQLConfigVersion
+	}
+
+	schema, err := graphql.NewSchemaFromString(g.config.Schema)
+	if err != nil {
+		return nil, err
+	}
+
+	conf := graphql.NewEngineV2Configuration(schema)
+
+	fieldConfigs, err := g.engineConfigV2FieldConfigs()
+	if err != nil {
+		return nil, err
+	}
+
+	datsSources, err := g.engineConfigV2DataSources()
+	if err != nil {
+		return nil, err
+	}
+
+	conf.SetFieldConfigurations(fieldConfigs)
+	conf.SetDataSources(datsSources)
+
+	return &conf, nil
+}
+
+func (g *GraphQLConfigAdapter) engineConfigV2FieldConfigs() (planFieldConfigs plan.FieldConfigurations, err error) {
+	for _, fc := range g.config.EngineConfig.FieldConfigs {
+		planFieldConfig := plan.FieldConfiguration{
+			TypeName:              fc.TypeName,
+			FieldName:             fc.FieldName,
+			DisableDefaultMapping: fc.DisableDefaultMapping,
+			Path:                  fc.Path,
+		}
+
+		planFieldConfigs = append(planFieldConfigs, planFieldConfig)
+	}
+
+	return planFieldConfigs, nil
+}
+
+func (g *GraphQLConfigAdapter) engineConfigV2DataSources() (planDataSources []plan.DataSourceConfiguration, err error) {
+	for _, ds := range g.config.EngineConfig.DataSources {
+		planDataSource := plan.DataSourceConfiguration{
+			RootNodes: []plan.TypeField{},
+		}
+
+		for _, typeField := range ds.RootFields {
+			planTypeField := plan.TypeField{
+				TypeName:   typeField.Type,
+				FieldNames: typeField.Fields,
+			}
+
+			planDataSource.RootNodes = append(planDataSource.RootNodes, planTypeField)
+		}
+
+		switch ds.Kind {
+		case GraphQLEngineDataSourceKindREST:
+			restConfig := GraphQLEngineDataSourceConfigREST{}
+			err = json.Unmarshal(ds.Config, &restConfig)
+			if err != nil {
+				return nil, err
+			}
+
+			factory := &rest_datasource.Factory{}
```

Comment:
`rest_datasource` is not a valid package call. Maybe `restDatasource`

Reply:
this package name is part of the library - I can only alias it in this file

Reply:
changed!